service/vector_store.cc

```diff
@@ -46,6 +63,117 @@ auto parse_service_uri(std::string_view uri) -> std::optional<std::tuple<host_na
     return {{host, *port}};
 }
 
+auto get_key_column_value(const rjson::value& item, std::size_t idx, const column_definition& column) -> std::expected<bytes, ann_error> {
+    auto const& column_name = column.name_as_text();
+    auto const* keys_obj = rjson::find(item, column_name);
+    if (keys_obj == nullptr) {
+        vslogger.error("Vector Store returned invalid JSON: missing key column '{}'", column_name);
+        return std::unexpected{service_reply_format_error{}};
+    }
+    if (!keys_obj->IsArray()) {
+        vslogger.error("Vector Store returned invalid JSON: key column '{}' is not an array", column_name);
+        return std::unexpected{service_reply_format_error{}};
+    }
+    auto const& keys_arr = keys_obj->GetArray();
+    if (keys_arr.Size() <= idx) {
+        vslogger.error("Vector Store returned invalid JSON: key column '{}' array too small", column_name);
+        return std::unexpected{service_reply_format_error{}};
+    }
+    auto const& key = keys_arr[idx];
+    return from_json_object(*column.type, key);
+}
+
+auto pk_from_json(rjson::value const& item, std::size_t idx, schema_ptr const& schema) -> std::expected<partition_key, ann_error> {
+    std::vector<bytes> raw_pk;
+    for (const column_definition& cdef : schema->partition_key_columns()) {
+        auto raw_value = get_key_column_value(item, idx, cdef);
+        if (!raw_value) {
+            return std::unexpected{raw_value.error()};
+        }
+        raw_pk.emplace_back(*raw_value);
+    }
+    return partition_key::from_exploded(raw_pk);
+}
+
+auto ck_from_json(rjson::value const& item, std::size_t idx, schema_ptr const& schema) -> std::expected<clustering_key_prefix, ann_error> {
+    if (schema->clustering_key_size() == 0) {
+        return clustering_key_prefix::make_empty();
+    }
+
+    std::vector<bytes> raw_ck;
+    for (const column_definition& cdef : schema->clustering_key_columns()) {
+        auto raw_value = get_key_column_value(item, idx, cdef);
+        if (!raw_value) {
+            return std::unexpected{raw_value.error()};
+        }
+        raw_ck.emplace_back(*raw_value);
+    }
+
+    return clustering_key_prefix::from_exploded(raw_ck);
+}
+
+auto write_ann_json(output_stream<char> out, embedding embedding, limit limit) -> future<> {
+    std::exception_ptr ex;
+    try {
+        co_await out.write(R"({"embedding":[)");
+        auto first = true;
+        for (auto value : embedding) {
+            if (!first) {
+                co_await out.write(",");
+            } else {
+                first = false;
+            }
+            co_await out.write(format("{}", value));
```

Comment:
Ugh, it's a pity that output_stream does not cooperate well with fmt::format... Hopefully, the small string optimization kicks in here and we will avoid an allocation.